server/src/main/java/org/elasticsearch/transport/TcpHeader.java

```diff
@@ -35,15 +35,39 @@
 
     public static final int VERSION_ID_SIZE = 4;
 
-    public static final int HEADER_SIZE = MARKER_BYTES_SIZE + MESSAGE_LENGTH_SIZE + REQUEST_ID_SIZE + STATUS_SIZE + VERSION_ID_SIZE;
+    public static final int VARIABLE_HEADER_SIZE = 4;
```

Comment:
Are we sure we don't just want to go with an unsigned short to have a "natural" `64k` limit on the header size? Seems to me we're just burning 2 bytes here (doesn't really matter with our protocol I guess) and opening ourselves up to abuse of the headers?

Reply:
I'm pretty indifferent on this topic. We don't use any unsigned types currently so this would be our first case of doing this. 

Looking around the internet I found that Tomcat defaults to max 8KB headers. Most servers have some max (although some servers limit the max to a single header value and some servers limit the max to total).

The case of whether we want to limit the total header size is kind of orthogonal to the defined header size. We can always place a defined limit.

The main question is if we want to do a short to save bytes and get 32KB. Or do a short AND add unsigned BE decoding to get 64KB.

@ywelsch

Reply:
> The case of whether we want to limit the total header size is kind of orthogonal to the defined header size. We can always place a defined limit.

Fair point. I'm definitely for limiting the header size overall. Having a defined maximum size for the header makes this whole exercies worth a lot more IMO (64k or less is kind of nice, as we would read that in one go in almost all setups and could then act on it accordingly).
Still, not sure if it's worth wasting extra bytes for an `int` here when we're placing a limit. Even though it doesn't matter much, we also won't need a spare 2 bytes IMO when we also have the variable size header in place anyway?

> The main question is if we want to do a short to save bytes and get 32KB. Or do a short AND add unsigned BE decoding to get 64KB.

I don't have a strong opinion here. It doesn't look like too much effort to read an unsigned short by just reading `2` bytes so I don't see a real downside to making the most of `2` bytes if we're gonna use `2`.

Reply:
I don't have a strong opinion here either. I'm fine with using a short here (32k), which is already way bigger than what many systems allow for headers (often 8k or 16k), or keep using the int, which does not have that much of an overhead (current header size is 19 bytes, and this would bring it to 21 or 23 respectively). I concur with Tim that we should avoid introducing a new way to encode/decode values.

Reply:
I'd say let's go for signed-short (32k) then :)

Reply:
Yannick and I looked through the code to identify where we use headers. Request headers are extracted from HTTP and sent internally based on what plugins require. Currently it seems extremely unlikely to ever be greater than 32KB.

In terms of response headers, our primary concern was that the Warning headers might be rather large. And by default we do not bound how many can be sent. Introducing a new transport limited header size seemed dangerous when the gains are only a 2 byte reduction. Additionally, although other systems currently have limits on header size, those are soft (configurable) limits. Whereas a transport defined limit requires a release + complicated BWC logic to mitigate.

I think it makes sense to use an int and we can introduce a configurable setting to limit total headers if we want to address the unbounded aspect.